### PR TITLE
Rework to library-go

### PR DESCRIPTION
**Pull request by jsafrane**
Rework to NewCSIControllerSet provided by library-go. In addition, following controllers are running:

* ManilaController
  * Checks if the cloud has Manila and starts CSIControllerSets for Manila and NFS.
  * Creates StorageClasses based on Manila shares
* SecretSync controller that translates cloud credentials to format understood by the driver.

The operator assumes that all connection details that the operator needs when talking to OpenStack is prepared by CSO, namely:
* CredentialRequest and its result in "manila-cloud-credentials" Secret.
* (Optional) self-signed CA certificate in "cloud-provider-config" ConfigMap.

@openshift/storage

### CODE REVIEW

File: pkg/controllers/manila/manila.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+	eventRecorder events.Recorder) factory.Controller {
+
+	scInformer := informers.InformersFor("").Storage().V1().StorageClasses()
+	c := &Controller{
+		operatorClient:     operatorClient,
+		kubeClient:         kubeClient,
+		storageClassLister: scInformer.Lister(),
+		openStackClient:    openStackClient,
+		csiControllers:     csiControllers,
+		eventRecorder:      eventRecorder,
```
Comment (bertinatto) [inline pkg/controllers/manila/manila.go:80]: prefix event recorder

File: pkg/controllers/manila/manila.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	shareTypes, err := c.openStackClient.GetShareTypes()
+	if err != nil {
+		switch err.(type) {
+		case *gophercloud.ErrEndpointNotFound:
+			// OpenStack does not support manila, report the operator as disabled
+			return c.setDisabled("This OpenStack does not provide Manila service")
```
Comment (bertinatto) [inline pkg/controllers/manila/manila.go:105]: OpenStack cluster?

File: pkg/controllers/manila/manila.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+		klog.V(4).Infof("Manila does not provide any share types")
+		return c.setDisabled("Manila does not provide any share types")
+	}
+	// Manila has some shares: start the actual CSI driver controller sets
+	if !c.controllersRunning {
+		klog.V(4).Infof("Starting CSI driver controllers")
+		for _, ctrl := range c.csiControllers {
+			go ctrl.Run(ctx, 1)
+		}
+		c.controllersRunning = true
```
Comment (bertinatto) [inline pkg/controllers/manila/manila.go:121]: nit: technically if `c.csiControllers` is empty this should be false

File: pkg/controllers/manila/manila.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+		sc := c.generateStorageClass(shareType)
+		err := c.applyStorageClass(ctx, sc)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) != 0 {
+		return errors.NewAggregate(errs)
+	}
+	return nil
```
Comment (bertinatto) [inline pkg/controllers/manila/manila.go:144]: I think you can just `return errors.NewAggregate(errs)`

File: test/e2e/storageclass.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: ebs.csi.aws.com
+provisioner: ebs.csi.aws.com
+volumeBindingMode: WaitForFirstConsumer
```
Comment (bertinatto) [inline test/e2e/storageclass.yaml:6]: you might want to remove these files, or the whole test dir

File: pkg/operator/starter.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+		kubeInformersForNamespaces,
+		controllerConfig.EventRecorder)
+
+	klog.Info("Starting the informers")
+	go kubeInformersForNamespaces.Start(ctx.Done())
+	go dynamicInformers.Start(ctx.Done())
+
+	klog.Info("Starting controllers")
+	go manilaController.Run(ctx, 1)
+	go secretSyncController.Run(ctx, 1)
```
Comment (bertinatto) [inline pkg/operator/starter.go:115]: This controller will run regardless of manila; can we move it to the `manila.Runnable` slice above and have it run only when manila is present?

File: pkg/controllers/manila/manila.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+			"type": shareType.Name,
+			"csi.storage.k8s.io/provisioner-secret-name":       util.ManilaSecretName,
+			"csi.storage.k8s.io/provisioner-secret-namespace":  util.OperatorNamespace,
+			"csi.storage.k8s.io/node-stage-secret-name":        util.ManilaSecretName,
+			"csi.storage.k8s.io/node-stage-secret-namespace":   util.OperatorNamespace,
+			"csi.storage.k8s.io/node-publish-secret-name":      util.ManilaSecretName,
+			"csi.storage.k8s.io/node-publish-secret-namespace": util.OperatorNamespace,
+		},
+	}
+	return sc
```
Comment (bertinatto) [inline pkg/controllers/manila/manila.go:189]: `storageclass-base.yaml` isn't being used

File: pkg/controllers/secret/secretsync.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	c := &Controller{
+		operatorClient: operatorClient,
+		kubeClient:     kubeClient,
+		secretLister:   secretInformer.Core().V1().Secrets().Lister(),
+		eventRecorder:  eventRecorder,
+	}
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).WithInformers(
+		operatorClient.Informer(),
+		secretInformer.Core().V1().Secrets().Informer(),
+	).ToController("SecretSync", eventRecorder)
```
Comment (bertinatto) [inline pkg/controllers/secret/secretsync.go:56]: There's no resync interval set with `ResyncEvery()` so this controller won't resync periodically.

File: pkg/controllers/secret/secretsync.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	secret := v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.ManilaSecretName,
+			Namespace: util.OperatorNamespace,
+		},
+		Type: v1.SecretTypeOpaque,
+		Data: data,
+	}
+
+	return &secret, nil
```
Comment (bertinatto) [inline pkg/controllers/secret/secretsync.go:158]: Side note: why does the CSI driver and the cluster-credentials-operator produce distinct secrets? Are both formats "officially" supported/valid in OpenStack? If so, I'd like to see either cloud-credentials-operator or the CSI driver supporting an additional format. But that's not a discussion for this PR...

File: assets/node-nfs.yaml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+          operator: Exists
+      containers:
+        - name: csi-driver
+          securityContext:
+            privileged: true
+          image: ${NFS_DRIVER_IMAGE}
+          args:
+            - "--nodeid=$(NODE_ID)"
+            - "--endpoint=unix://plugin/csi.sock"
+            - "--mount-permissions=0777"
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/node-nfs.yaml:30]: Yes! There is no fsgroup on NFS and this makes the volume accessible to anyone.

File: cmd/csi-driver-manila-operator/main.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
+package main
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply cmd/csi-driver-manila-operator/main.go:1]: I'm not sure, this is driver manifest, not the operator.

File: pkg/controllers/manila/manila.go (hunk 3 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go:121]: they can't be empty...

File: pkg/controllers/secret/secretsync.go (hunk 9 above)
Comment (jsafrane) [reply pkg/controllers/secret/secretsync.go:158]: I talked to @Fedosin and there are at least three different confing/secret formats in OpenStack, each component using its own. I agree it should be in cloud-credentials-operator

File: pkg/controllers/secret/secretsync.go (hunk 8 above)
Comment (jsafrane) [reply pkg/controllers/secret/secretsync.go:56]: Added 20 minutes resync

File: pkg/controllers/manila/manila.go (hunk 7 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go:189]: Removed

File: pkg/operator/starter.go (hunk 6 above)
Comment (jsafrane) [reply pkg/operator/starter.go]: Fixed

File: test/e2e/storageclass.yaml (hunk 5 above)
Comment (jsafrane) [reply test/e2e/storageclass.yaml]: Removed whole test directory

File: pkg/controllers/manila/manila.go (hunk 4 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go]: Fixed

File: pkg/controllers/manila/manila.go (hunk 2 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go]: Fixed

File: pkg/controllers/manila/manila.go (hunk 1 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go]: Fixed

File: pkg/operator/starter.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+	nfsController := csidrivercontroller.NewCSIDriverController(
+		"NFSDriverController",
+		string(operatorapi.ManilaCSIDriver),
+		"nfs-csi-driver",
+		util.OperatorNamespace,
+		operatorClient,
+		assetWithNFSDriver,
+		kubeClient,
+		controllerConfig.EventRecorder,
+	).WithNodeService(kubeInformersForNamespaces.InformersFor(util.OperatorNamespace).Apps().V1().DaemonSets(), "node-nfs.yaml")
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/operator/starter.go]: Fixed

File: pkg/operator/starter.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+)
+
+const (
+	operandName = "manila-csi-driver"
+
+	nfsImageEnvName = "NFS_DRIVER_IMAGE"
+)
+
+func RunOperator(ctx context.Context, controllerConfig *controllercmd.ControllerContext) error {
+	kubeClient := kubeclient.NewForConfigOrDie(rest.AddUserAgent(controllerConfig.KubeConfig, "kube-client"))
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/operator/starter.go]: Used operator name

File: assets/controller.yaml [hunk 14]
```diff
@@ -0,0 +1,133 @@
+kind: Deployment
+apiVersion: apps/v1
+metadata:
+  name: manila-csi-driver-controller
+  namespace: openshift-cluster-csi-drivers
+spec:
+  selector:
+    matchLabels:
+      app: manila-csi-driver-controller
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/controller.yaml:9]: same as above, it's driver, not operator

File: pkg/dependencymagnet/dependencymagnet.go [hunk 15]
```diff
@@ -0,0 +1,11 @@
+// +build tools
+
+// go mod won't pull in code that isn't depended upon, but we have some code we don't depend on from code that must be included
+// for our build to work.
+package dependencymagnet
+
+import (
+	_ "github.com/go-bindata/go-bindata/go-bindata"
+	_ "github.com/openshift/build-machinery-go"
+	_ "k8s.io/code-generator"
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/dependencymagnet/dependencymagnet.go]: Fixed

File: go.mod [hunk 16]
```diff
@@ -3,23 +3,22 @@ module github.com/openshift/csi-driver-manila-operator
...
-replace (
-	github.com/Azure/go-autorest => github.com/Azure/go-autorest v13.3.2+incompatible // Required by OLM
-	k8s.io/client-go => k8s.io/client-go v0.17.4 // Required by prometheus-operator
+	gopkg.in/yaml.v2 v2.3.0
+	k8s.io/api v0.19.0-rc.2
+	k8s.io/apimachinery v0.19.0-rc.2
+	k8s.io/client-go v0.19.0-rc.2
+	k8s.io/code-generator v0.19.0-rc.2
+	k8s.io/component-base v0.19.0-rc.2
+	k8s.io/klog v1.0.0
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply go.mod]: Fixed

File: assets/node-nfs.yaml [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+            - name: NODE_ID
+              valueFrom:
+                fieldRef:
+                  fieldPath: spec.nodeName
+          volumeMounts:
+            - name: plugin-dir
+              mountPath: /plugin
+            - name: pods-mount-dir
+              mountPath: /var/lib/kubelet/pods
+              mountPropagation: Bidirectional
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/node-nfs.yaml]: Fixed

File: assets/controller.yaml [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+            requests:
+              cpu: 10m
+              memory: 50Mi
+        # TODO: fix manila CSI driver not to require NFS driver socket!
+        - name: csi-driver-nfs
+          image: ${NFS_DRIVER_IMAGE}
+          args:
+            - "--nodeid=$(NODE_ID)"
+            - "--endpoint=unix://plugin/csi-nfs.sock"
+            - "--mount-permissions=0777"
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/controller.yaml:68]: Fixed

File: assets/controller.yaml [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+              mountPath: /var/lib/csi/sockets/pluginproxy/
+          resources:
+            requests:
+              cpu: 10m
+              memory: 50Mi
+        - name: csi-snapshotter
+          securityContext:
+            privileged: true
+            capabilities:
+              add: ["SYS_ADMIN"]
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/controller.yaml]: No! Removed privileged from controller containers

File: assets/controller.yaml [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+        # Warning: the operator expects the first container to be the CSI driver
+        - name: csi-driver
+          securityContext:
+            privileged: true
+            capabilities:
+              add: ["SYS_ADMIN"]
+            allowPrivilegeEscalation: true
+          image: ${DRIVER_IMAGE}
+          args:
+            - --v=5
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/controller.yaml]: Fixed

File: assets/controller.yaml [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+      labels:
+        app: manila-csi-driver-controller
+    spec:
+      serviceAccount: manila-csi-driver-controller-sa
+      priorityClassName: system-cluster-critical
+      tolerations:
+        - key: CriticalAddonsOnly
+          operator: Exists
+      containers:
+        # Warning: the operator expects the first container to be the CSI driver
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/controller.yaml]: Fixed

File: Makefile [hunk 22]
```diff
@@ -0,0 +1,54 @@
+SHELL :=/bin/bash
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply Makefile]: Fixed

File: Makefile [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+IMAGE_REGISTRY?=registry.svc.ci.openshift.org
+
+# This will call a macro called "build-image" which will generate image specific targets based on the parameters:
+# $0 - macro name
+# $1 - target name
+# $2 - image ref
+# $3 - Dockerfile path
+# $4 - context directory for image build
+# It will generate target "image-$(1)" for building the image and binding it as a prerequisite to target "images".
+$(call build-image,csi-driver-manila-operator,$(IMAGE_REGISTRY)/ocp/4.5:csi-driver-manila-operator,./build/Dockerfile.openshift,.)
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply Makefile]: Fixed

File: cmd/csi-driver-manila-operator/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (bertinatto) [reply cmd/csi-driver-manila-operator/main.go:1]: I'm not referring to the manifest file, I'm talking about `cmd/csi-driver-manila-operator` directory

File: pkg/operator/starter.go [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+	csicontrollerset "github.com/openshift/library-go/pkg/operator/csi/csicontrollerset"
+	csidrivercontroller "github.com/openshift/library-go/pkg/operator/csi/csidrivercontroller"
+	goc "github.com/openshift/library-go/pkg/operator/genericoperatorclient"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	operandName  = "manila-csi-driver"
+	operatorName = "manila-csi-driver-operator"
```
Comment (bertinatto) [inline pkg/operator/starter.go:28]: note: the operator namespace is in `util`, not sure if want to have them in the same place

File: pkg/operator/starter.go [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+
+	nfsController := csidrivercontroller.NewCSIDriverController(
+		"NFSDriverController",
+		string(operatorapi.ManilaCSIDriver),
+		"nfs-csi-driver",
+		util.OperatorNamespace,
+		operatorClient,
+		assetWithNFSDriver,
+		kubeClient,
+		controllerConfig.EventRecorder,
```
Comment (bertinatto) [inline pkg/operator/starter.go:89]: The order of these args is incorrect:
```
func NewCSIDriverController(
	name string,
	csiDriverName string,
	csiDriverNamespace string,
	instanceName string,
	operatorClient v1helpers.OperatorClient,
	assetFunc func(string) []byte,
	kubeClient kubernetes.Interface,
	eventRecorder events.Recorder,
```

File: pkg/controllers/manila/manila.go [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+
+type Runnable interface {
+	Run(ctx context.Context, workers int)
+}
+
+const (
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Manila and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 1 * time.Minute
```
Comment (bertinatto) [inline pkg/controllers/manila/manila.go:60]: Do we need to resync every minute?

File: go.mod [hunk 27]
```diff
@@ -3,23 +3,21 @@ module github.com/openshift/csi-driver-manila-operator
...
-replace (
-	github.com/Azure/go-autorest => github.com/Azure/go-autorest v13.3.2+incompatible // Required by OLM
-	k8s.io/client-go => k8s.io/client-go v0.17.4 // Required by prometheus-operator
+	gopkg.in/yaml.v2 v2.3.0
+	k8s.io/api v0.19.0-rc.2
+	k8s.io/apimachinery v0.19.0-rc.2
+	k8s.io/client-go v0.19.0-rc.2
+	k8s.io/component-base v0.19.0-rc.2
+	k8s.io/klog/v2 v2.3.0
+	sigs.k8s.io/yaml v1.2.0
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply go.mod:22]: `go mod tidy` does that.

File: pkg/operator/starter.go [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+		resync,
+		controllerConfig.EventRecorder)
+
+	manilaController := manila.NewController(
+		operatorClient,
+		kubeClient,
+		kubeInformersForNamespaces,
+		openstackClient,
+		[]manila.Runnable{
+			manilaControllerSet,
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/operator/starter.go:110]: Renamed

File: pkg/operator/starter.go (hunk 24 above)
Comment (jsafrane) [reply pkg/operator/starter.go:28]: It's name of the operator, not name of the namespace. It's just coincidence they're the same :-)

File: pkg/controllers/manila/manila.go [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+// 2) Creates StorageClass for each share type provided by Manila.
+// 3) If there is no Manila in the OpenStack where the cluster runs,
+//    it marks the operator with condition Disabled=true.
+//
+// Note that the CSI driver(s) are not un-installed when Manila becomes
+// missing or it stops providing shares of given type - Manila bight be
+// under (short?) maintenance / reconfiguration.
+// Similarly, StorageClasses are not deleted when a share type disappears
+// from Manila.
+type Controller struct {
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/controllers/manila/manila.go]: Renamed

File: pkg/controllers/manila/manila.go [hunk 30]
```diff
@@ -0,0 +1,218 @@
...
+		if !reflect.DeepEqual(expected.Parameters, current.Parameters) {
+			// StorageClass.Parameters changed. Typically, secret namespace
+			// is different when moving from OLM to non-OLM operator.
+			// Delete the old class and create a new one.
+			if err := c.kubeClient.StorageV1().StorageClasses().Delete(ctx, expected.Name, metav1.DeleteOptions{}); err != nil {
+				if apierrors.IsNotFound(err) {
+					err = nil
+				}
+				return err
+			}
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/controllers/manila/manila.go:156]: Fixed + added comment about fallthrough

File: pkg/controllers/manila/manila.go [hunk 31]
```diff
@@ -0,0 +1,218 @@
...
+		Type:   operatorConditionPrefix + operatorv1.OperatorStatusTypeAvailable,
+		Status: operatorv1.ConditionTrue,
+	}
+	_, _, err := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(availableCnd),
+		removeConditionFn(operatorConditionPrefix+"Disabled"))
+	return err
+}
+
+func (c *Controller) setDisabled(msg string) error {
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/controllers/manila/manila.go]: Renamed

File: pkg/operator/starter.go (hunk 25 above)
Comment (jsafrane) [reply pkg/operator/starter.go:89]: Nice catch, fixed

File: pkg/controllers/manila/manila.go (hunk 26 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go:60]: When someone creates a new share type in OpenStack, this operator should create StorageClass for it. Question is what's the right frequency.

File: pkg/operator/starter.go (hunk 24 above)
Comment (bertinatto) [reply pkg/operator/starter.go:28]: I meant to say: have the operator name and namespace consts in the same place. But that's not really important

File: pkg/controllers/manila/manila.go (hunk 26 above)
Comment (bertinatto) [reply pkg/controllers/manila/manila.go:60]: IIUC the optimal resync interval for this type of storage backend (manila, EFS) will vary from cluster to cluster. For instance, in some clusters the number of share types will seldom change, while in others new share types will be added and removed quite often.

Having said that, whatever we hardcode here might not be the ideal for some use cases.

IMO the best way to approach this is to define a standard value, say 20 minutes, and allow the cluster admin to customize it through the CR. If we choose this approach, we could tackle the customization part in another PR. 

Thoughts?

File: README.md [hunk 32]
```diff
@@ -1,112 +1,26 @@
-# CSI Driver Manila Operator
+# Manila CSI driver operator
 
-Operator to create, configure and manage CSI driver for OpenStack Manila in OpenShift.
+An operator to deploy the [Manil CSI driver](https://github.com/openshift/cloud-provider-openstack/tree/master/pkg/csi/manila) in OpenShift.
```
Comment (bertinatto) [inline README.md:3]: Manil**a**

File: README.md [hunk 33]
```diff
@@ -1,112 +1,26 @@
...
 
-## Quick Start
+## Design
 
-### Installing the operator
+The operator is based on [openshift/library-go](https://github.com/openshift/library-go). It manages `ManilaDriver` instance named `cluster` and runs several controllers in parallel:
```
Comment (bertinatto) [inline README.md:7]: Kind and CR changed

File: pkg/controllers/manila/manila.go (hunk 26 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go]: I'm updating it to 20 min, however, I am not sure how important it is to have it customizable.

File: pkg/operator/starter.go [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+			"rbac/provisioner_binding.yaml",
+			"rbac/provisioner_role.yaml",
+			"rbac/privileged_role.yaml",
+			"rbac/controller_privileged_binding.yaml",
+			"rbac/node_privileged_binding.yaml",
+		},
+	).WithCSIDriverController(
+		"ManilaDriverController",
+		operandName,
+		string(operatorapi.ManilaCSIDriver),
```
Comment (bertinatto) [inline pkg/operator/starter.go:72]: Aren't the instance/config name and driver/operand name swapped here?

File: pkg/operator/starter.go [hunk 35]
```diff
@@ -0,0 +1,140 @@
...
+		kubeInformersForNamespaces.InformersFor(util.OperatorNamespace),
+		csicontrollerset.WithControllerService("controller.yaml"),
+		csicontrollerset.WithNodeService("node.yaml"),
+	)
+
+	nfsCSIDriverController := csidrivercontroller.NewCSIDriverController(
+		"NFSDriverController",
+		string(operatorapi.ManilaCSIDriver),
+		util.OperatorNamespace,
+		"nfs-csi-driver",
```
Comment (bertinatto) [inline pkg/operator/starter.go:85]: Swap line 83 with line 85?

File: README.md [hunk 36]
```diff
@@ -1,112 +1,26 @@
...
 
-The operator needs its own namespace, service account, security context, and a few roles and bindings. For example, to install these on OpenShift >= 4.4:
+* `manilaController`: Talks to OpenStack API and checks if Manila service is provided.
+  * If Manila service is found: 
+    * It starts `manilaControllerSet`: Runs `csidriverset.Controller` that installs the Manila CSI driver itself.
+    * It starts `nfsController`: Runs `csidriverset.Controller` that installs NFS CSI driver itself.
+    * It creates `StorageClass` for each share type reported by Manila and periodically syncs them at least once per minute, in case a new share type appears in Manila.
+    * It never removes StorageClass if Manila service or share type disappears. It may be temporary OpenStack ir Manila re-configuration hiccup.
+  * If there is no Manila service, it marks the `ManilaDriver` instance with `ManilaControllerDisabled: True` condition. It does not stop any CSI drivers started when Manila service was present! This allows pod to at least unmount their volumes.
```
Comment (bertinatto) [inline README.md:15]: Replace ManilaDriver with ClusterCSIDriver

File: assets/controller.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (bertinatto) [reply assets/controller.yaml:9]: Sorry, meant `openstack-manila-csi-driver-controller`, as in `aws-ebs-csi-driver-controller`.

Prefixing the cloud platform is useful in AWS, Azure, GCE etc., not sure if it's useful for OpenStack (manila and cinder are kind of unique).

File: pkg/controllers/manila/manila.go [hunk 37]
```diff
@@ -0,0 +1,218 @@
...
+//
+// Note that the CSI driver(s) are not un-installed when Manila becomes
+// missing or it stops providing shares of given type - Manila bight be
+// under (short?) maintenance / reconfiguration.
+// Similarly, StorageClasses are not deleted when a share type disappears
+// from Manila.
+type ManilaController struct {
+	operatorClient     v1helpers.OperatorClient
+	kubeClient         kubernetes.Interface
+	secretLister       corelisters.SecretLister
```
Comment (bertinatto) [inline pkg/controllers/manila/manila.go:43]: I think you're not using this anymore

File: pkg/controllers/manila/openstack.go [hunk 38]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	provider, err := openstack.NewClient(opts.IdentityEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	cert, err := getCloudProviderCert()
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("Failed to get cloud provider CA certificate: %v", err)
```
Comment (bertinatto) [inline pkg/controllers/manila/openstack.go:59]: nit: my editor says errors should not start with capital letters

File: pkg/controllers/manila/openstack.go [hunk 39]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	if len(cert) > 0 {
+		certPool, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, fmt.Errorf("Create system cert pool failed: %v", err)
```
Comment (bertinatto) [inline pkg/controllers/manila/openstack.go:65]: ditto

File: README.md (hunk 32 above)
Comment (jsafrane) [reply README.md:3]: Fixed

File: README.md (hunk 33 above)
Comment (jsafrane) [reply README.md:7]: Fixed

File: pkg/operator/starter.go (hunk 34 above)
Comment (jsafrane) [reply pkg/operator/starter.go:72]: Fixed, I think

File: pkg/operator/starter.go (hunk 35 above)
Comment (jsafrane) [reply pkg/operator/starter.go:85]: Fixed, I think

File: README.md (hunk 36 above)
Comment (jsafrane) [reply README.md:15]: Fixed

File: pkg/controllers/manila/manila.go (hunk 37 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go:43]: Removed

**Comment by bertinatto** [comment]
/lgtm

Adding hold in case @Fedosin wants to take a look
/hold

File: README.md [hunk 40]
```diff
@@ -1,112 +1,26 @@
...
-```sh
-oc delete -f deploy/crds/csi.openshift.io_v1alpha1_maniladriver_cr.yaml
-```
-
-When the driver is deleted, remove the remaining parts of the operator.
-
-```sh
-oc delete -f deploy/crds/csi.openshift.io_maniladrivers_crd.yaml -f deploy/role.yaml -f deploy/role_binding.yaml -f deploy/service_account.yaml -f deploy/namespace.yaml
-```
+* OpenStack cloud credentials are in Secret named "cloud-credentials" in the same namespace where the operator runs. The operator uses the credentials to check if Manila is present in the cluster and, since OpenStack does not allow any fine-grained access controll, it lets the CSI driver to use the same credentials.
```
Comment (Fedosin) [inline README.md:24]: access control

File: README.md [hunk 41]
```diff
@@ -1,112 +1,26 @@
...
+* If underlying OpenStack uses self-signed certificate, the operator expects the ceritificate is present in a ConfigMap named "cloud-provider-config" with key "ca-bundle.pem" in the namespace where it runs. Generally, it should be a copy of "openshift-config/cloud-provider-config" ConfigMap. It then uses the certificate to talk to OpenStack API.
```
Comment (Fedosin) [inline README.md:25]: certificate

File: pkg/version/version.go [hunk 42]
```diff
@@ -0,0 +1,47 @@
...
+		GitCommit:  commitFromGit,
+		GitVersion: versionFromGit,
+		BuildDate:  buildDate,
+	}
+}
+
+func init() {
+	buildInfo := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "openshift_aws_ebs_csi_driver_operator",
```
Comment (Fedosin) [inline pkg/version/version.go:39]: openshift_manila_csi_driver_operator

File: pkg/version/version.go [hunk 43]
```diff
@@ -0,0 +1,47 @@
...
+			Help: "A metric with a constant '1' value labeled by major, minor, git commit & git version from which OpenShift AWS EBS CSI Driver Operator was built.",
```
Comment (Fedosin) [inline pkg/version/version.go:40]: Openshift Manila CSI Driver

File: README.md (hunk 40 above)
Comment (jsafrane) [reply README.md:24]: fixed

File: README.md (hunk 41 above)
Comment (jsafrane) [reply README.md]: fixed

File: pkg/version/version.go (hunk 42 above)
Comment (jsafrane) [reply pkg/version/version.go]: good catch, fixed

File: pkg/version/version.go (hunk 43 above)
Comment (jsafrane) [reply pkg/version/version.go]: fixed

**Pull request merged**